RecoBTag/ONNXRuntime/plugins/ParticleNetSonicJetTagsProducer.cc

```diff
@@ -0,0 +1,317 @@
+      std::cout << " - Jet #" << jet_n << ", pt=" << jet_ref->pt() << ", eta=" << jet_ref->eta()
+                << ", phi=" << jet_ref->phi() << std::endl;
+      for (std::size_t flav_n = 0; flav_n < flav_names_.size(); ++flav_n) {
+        std::cout << "    " << flav_names_.at(flav_n) << " = " << (*(output_tags.at(flav_n)))[jet_ref] << std::endl;
+      }
+    }
+  }
+
+  // put into the event
+  for (std::size_t flav_n = 0; flav_n < flav_names_.size(); ++flav_n) {
+    iEvent.put(std::move(output_tags[flav_n]), flav_names_[flav_n]);
+  }
+}
+
+std::vector<float> ParticleNetSonicJetTagsProducer::center_norm_pad(const std::vector<float> &input,
+                                                                    float center,
+                                                                    float norm_factor,
+                                                                    unsigned min_length,
+                                                                    unsigned max_length,
+                                                                    float pad_value,
+                                                                    float replace_inf_value,
+                                                                    float min,
+                                                                    float max) {
+  // do variable shifting/scaling/padding/clipping in one go
+
+  assert(min <= pad_value && pad_value <= max);
+  assert(min_length <= max_length);
+
+  unsigned target_length = std::clamp((unsigned)input.size(), min_length, max_length);
+  std::vector<float> out(target_length, pad_value);
+  for (unsigned i = 0; i < input.size() && i < target_length; ++i) {
+    out[i] = std::clamp((catch_infs(input[i], replace_inf_value) - center) * norm_factor, min, max);
+  }
+  return out;
+}
+
+void ParticleNetSonicJetTagsProducer::acquire(edm::Event const &iEvent, edm::EventSetup const &iSetup, Input &iInput) {
+  client_->setBatchSize(batchSize_);
+  edm::Handle<TagInfoCollection> tag_infos;
+  iEvent.getByToken(src_, tag_infos);
+  if (!tag_infos->empty()) {
+    unsigned jet_n = 0;
+    const auto &taginfo = (*tag_infos)[jet_n];
+    for (unsigned igroup = 0; igroup < input_names_.size(); ++igroup) {
+      const auto &group_name = input_names_[igroup];
+      auto &input = iInput.at(group_name);
+      auto tdata = std::make_shared<TritonInput<float>>(1);
+      auto &vdata = (*tdata)[0];
+      vdata.reserve(input.sizeShape());
+
+      const auto &prep_params = prep_info_map_.at(group_name);
+      // first reset group_values to 0
+      std::fill(vdata.begin(), vdata.end(), 0);
+      unsigned curr_pos = 0;
+      // transform/pad
+      for (unsigned i = 0; i < prep_params.var_names.size(); ++i) {
+        const auto &varname = prep_params.var_names[i];
+        const auto &raw_value = taginfo.features().get(varname);
+        const auto &info = prep_params.info(varname);
+        auto val = center_norm_pad(raw_value,
```

Comment:
it would be better to pass `vdata()` to `center_norm_pad` and fill it directly (avoids unnecessary copies/fills)

Reply:
Done, but please let me know if you think I need to improve the way I implemented it.  The center_norm_pad function is now a helper in deep_helpers